### PR TITLE
Feature/block content.type

### PR DESCRIPTION
**Pull request by d34dman**
Apologies for, 
- PHP code sniffer not installed
- the branch name
- the super mega merge

TLDR;
- Implements scaffolding for block content type
- Add support for following fields on block content type
  - image
  - text
  - textarea
  - cta
- Implements scaffolding for Responsive Image

Long Description:

- Decoupled the logic of template crunching and exposing the transformer for phabalicious
- Which means we can add and re-use a transformer for a given config without exposing it to phabalicious. Please take a peak into "src/transformers/utils" folder.
- Added a demo folder in root, which can be used as reference and maybe for tests in future.

### CODE REVIEW

**Comment by stmh** [comment]
I made some changes which need the latest version of phabalicious:

* I removed the require_once and made sure namespaces are registered globally on phabablicious side.
* phab exposes the target_path to the transformers, so they can read existing data if necessary

Changes are pushed to both repos

Some minor suggestions:
* I'd rename ConfigService to sth different, as it is not a service per se. Also there is an existing ConfigurationService in phabalicious which might confuse users. This should be named sth so it is more clear that this object contains the actual result. From the naming I thougth first it holds some needed configuration.
* personally I am not a big fan to do all the computation in constructors, as this makes it hard to follow the program flow

I did some real world tests:
block_content is empty when scaffolding it, and the configuration cant be imported:
> Attempt to create a field field_card_headline that does not exist on entity type block_content.

**Comment by d34dman** [comment]
> I made some changes which need the latest version of phabalicious:
> 
> * I removed the require_once and made sure namespaces are registered globally on phabablicious side.
> * phab exposes the target_path to the transformers, so they can read existing data if necessary
> 
> Changes are pushed to both repos
> 

Reply: For some reason I couldn't get the autoloader working on my local anymore. Am using this commit of phabalicious fa19e400f3b1ef502bcf7e0065d95960f39e908a with bin/phab hacked to report version as 3.4




> Some minor suggestions:
> 
> * I'd rename ConfigService to sth different, as it is not a service per se. Also there is an existing ConfigurationService in phabalicious which might confuse users. This should be named sth so it is more clear that this object contains the actual result. From the naming I thougth first it holds some needed configuration.

Reply: Changed name to ConfigAccumulator, I would be more than happy to add developer documentation once we finalise on these things :)


> * personally I am not a big fan to do all the computation in constructors, as this makes it hard to follow the program flow

Reply: Am open for suggestions :) 


> 
> I did some real world tests:
> block_content is empty when scaffolding it, and the configuration cant be imported:

Reply: I am sorry, must confess I did a lot of refactoring in last couple of days. I would try to come up with a working solution once the autoloader is fixed. (I can't test it on my local anymore). Getting following output :(

```
phab scaffold es/index.yml                                                                         1.2m  Mon Mar  2 23:11:44 2020


 Version 3.3.6 of phabalicious is available. Run `phab self-update` to update your local installation.

 Visit https://github.com/factorial-io/phabalicious/releases for more info.



 // Create destination folder ...

 // Start scaffolding script ...

 // Transforming responsive_image ...


Fatal error: Uncaught Error: Class 'Phabalicious\Scaffolder\Transformers\Utils\PlaceholderService' not found in /Users/shibindas/work/multibasebox/projects/factorial-website/vendor/factorial-io/phab-entity-scaffolder/src/transformers/ResponsiveImageTransformer.php:26
Stack trace:
#0 phar:///usr/local/bin/phab/src/Scaffolder/Callbacks/TransformCallback.php(75): Phabalicious\Scaffolder\Transformers\ResponsiveImageTransformer->transform(Object(Phabalicious\Method\TaskContext), Array, '/Users/shibinda...')
#1 phar:///usr/local/bin/phab/src/Scaffolder/Callbacks/TransformCallback.php(40): Phabalicious\Scaffolder\Callbacks\TransformCallback->transform(Object(Phabalicious\Method\TaskContext), 'responsive_imag...', 'responsive_imag...', 'config/sync')
#2 [internal function]: Phabalicious\Scaffolder\Callbacks\TransformCallback->handle(Object(Phabalicious\Method\TaskContext), 'responsive_imag...', 'responsive_imag...', 'config/sync')
#3 phar:///usr/local/bin/phab/src/Method/ScriptMethod.php(236): call_user_func_array(Array, Array)
# in /Users/shibindas/work/multibasebox/projects/factorial-website/vendor/factorial-io/phab-entity-scaffolder/src/transformers/ResponsiveImageTransformer.php on line 26
```


> > Attempt to create a field field_card_headline that does not exist on entity type block_content.

Reply: Will try to fix these once i am able to run phab in my local. Probably also add some tests.

File: src/transformers/Utils/ConfigService.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    public function __construct()
+    {
+        $this->config = [];
+    }
+
+    public function get() {
+        $config = $this->getRaw();
+        $output = [];
+        foreach($config as $key => $data) {
+            if (!empty($data['dependenices'])) {
```
[the remark this answers is not in the document]
Comment (d34dman) [reply src/transformers/Utils/ConfigService.php]: Fixed

File: src/transformers/Utils/ConfigService.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    public function setConfig($config_name, $value) {
+        $this->config[$config_name] = $value;
+    }
+
+    public function getConfig($config_name) {
+        return $this->config[$config_name];
+    }
+
+    public function getConfigUUID($config_name) {
```
[the remark this answers is not in the document]
Comment (d34dman) [reply src/transformers/Utils/ConfigService.php]: It's not being used now, but is would play an important role in the future development.

File: src/transformers/Utils/EntityBase.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+use Phabalicious\Scaffolder\Transformers\Utils\PlaceholderService;
+use Phabalicious\Scaffolder\Transformers\Utils\Base;
+use Phabalicious\Scaffolder\Transformers\Utils\EntityForm;
+use Phabalicious\Scaffolder\Transformers\Utils\FieldField;
+use Phabalicious\Scaffolder\Transformers\Utils\FieldStorage;
+use Phabalicious\Scaffolder\Transformers\Utils\FieldWidget;
+
+abstract class EntityBase extends Base
+{
+    const ENTITY_TYPE = '???';
```
[the remark this answers is not in the document]
Comment (d34dman) [reply src/transformers/Utils/EntityBase.php]: Done :) we have a function called getEntityType now ):

File: src/transformers/Utils/EntityPropertyBase.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
+<?php
```
[the remark this answers is not in the document]
Comment (d34dman) [reply src/transformers/Utils/EntityPropertyBase.php]: Yes very much. The classes that extends these are Fields and Image Effects which can be visualised as property of an object like "Entity" or Image styles.

**Comment by d34dman** [comment]
And thanks a lot for taking time to review, very much appreciated :)

**Comment by stmh** [comment]
Are you sure you are using the right phabalicious version? As the used commit is using version 3.4 so no patching needed. I'd propose writing unit tests for the scaffolders as they wil use the right phab version out of the box. and impove DX a lot

**Comment by d34dman** [comment]
@stmh , thanks for the review and guidance! much appreciated!

- Added tests
    - block_content
    - responsive image
    - nova's ark (for testing all fields)
- for responsive images, the fallback image style is populated now.
- The tests works more or less like a regression test at the moment, i.e. i copied the output generated into base line :P.
